### ISSUE TITLE
Rails uniqueness validator incorrectly checks previous record.

### ISSUE THREAD

**Issue by pharmmd-rich** [rails/rails#23399]
The Rails uniqueness validator performs it's query based on the current value of the id attribute as opposed to the original value of id (id_was).  This leads to an error condition when you allow the user to update the id itself.  

Example:

Migration: 

```
class CreateOrganizations < ActiveRecord::Migration
  def change
    create_table :organizations, id: false do |t|
      t.integer :id, primary_key: true
      t.string :name

      t.timestamps null: false
    end
  end
end
```

Model:

```
class Organization < ActiveRecord::Base
  validates :id, presence: true, uniqueness: true
  validates :name, presence: true, uniqueness: true
end
```

Code:

```
o1 = Organization.create!(id: 1, name: "Org 1")
o2 = Organization.create!(id: 2, name: "Org 2")

o1.id = o2.id

o1.save!
```

In the above example, you should get a validation error on ID, instead you get it on name because Rails generates the query based on the newly assigned id of 2 instead of the original id of 1

`SELECT  1 AS one FROM "organizations" WHERE ("organizations"."name" = 'Org 1' AND "organizations"."id" != 2) LIMIT 1`

Error reproduced successfully in 4.2.5, haven't tried 5.0 beta yet.

**Comment by maclover7** [comment]
I think this might be an error within your reproduction script -- I think you mean to be calling `o1.save!` in the last line of your script above. A properly formatted script can be found here -- https://gist.github.com/maclover7/7fe5b3d47332d9d5458c. I'd say there's no unexpected behavior here.

**Comment by pharmmd-rich** [comment]
Just edited, result is the same.

**Comment by maclover7** [comment]
My script posted above seems like it's demonstrating the expect behavior -- do you also see this?

**Comment by pharmmd-rich** [comment]
Your test incorrectly tests for ActiveRecord::RecordInvalid.  ActiveRecord::RecordInvalid is thrown, but on the incorrect column of name rather than id.  Incorrect tests cause incorrect results!  Watch those tests ;)

**Comment by maclover7** [comment]
If you look at the error log for `ActiveRecord::RecordInvalid`, you'd see that the error message was that both of the columns were invalid, and could not be saved because of the model's listed validation requirements.

**Comment by betam4x** [comment]
@maclover7 incorrect.  The error log states nothing about ID, and why would the name be invalid at all?  Did you even look at your own test?  Even the rails source code says it uses the primary key, which if edited, would be different from what was persisted:

**/activerecord/lib/active_record/validations/uniqueness.rb:23**
`relation = relation.and(table[finder_class.primary_key.to_sym].not_eq(record.send(:id))) if record.persisted?`

in this case that translates to `id != 'id'`.  however the `id` was modified in this case (in the case above, id, is changed to 2 for org1, so the query uses 2 instead of 1 to exclude itself from the check, resulting in unpredictable behavior), so you you end up with incorrect behavior.

Note that the primary key could be ANYTHING in this case and the behavior would still be the same:  updates involving the primary key will always be incorrect when the uniqueness validator is used (`organization_id != 2` instead of `organization_id != 1` for instance)

Reproduced on Rails 4.2.5.1 on Windows 10, Ruby 2.2.3p173 and SQLite3

**Comment by kamipo** [comment]
Fixed in #23581.
